cassandane/tiny-tests/Sieve/jmapquery_multiple_deliveredto

```diff
@@ -0,0 +1,78 @@
+#!perl
+use Cassandane::Tiny;
+
+sub test_jmapquery_multiple_deliveredto
+    :min_version_3_3 :needs_component_jmap
+{
+    my ($self) = @_;
+
+    my $imap = $self->{store}->get_client();
+    $imap->create("INBOX.matches") or die;
+
+    $self->{instance}->install_sieve_script(<<'EOF'
+require ["x-cyrus-jmapquery", "x-cyrus-log", "variables", "fileinto"];
+if
+  allof( not string :is "${stop}" "Y",
+    jmapquery text:
+  {
+    "operator" : "OR",
+    "conditions" : [
+        {
+           "deliveredTo" : "xxx@yyy.zzz"
+        },
+        {
+           "deliveredTo" : "xxx+*@yyy.zzz"
+        }
+    ],
+    "operator" : "OR"
+  }
+.
+  )
+{
+  fileinto "INBOX.matches";
+}
+EOF
+    );
+
+    my $body = << 'EOF';
+--047d7b33dd729737fe04d3bde348
+Content-Type: text/plain; charset=UTF-8
+
+plain
+
+--047d7b33dd729737fe04d3bde348
+Content-Type: image/tiff
+Content-Transfer-Encoding: base64
+
+abc=
+
+--047d7b33dd729737fe04d3bde348--
+EOF
+    $body =~ s/\r?\n/\r\n/gs;
+
+    xlog $self, "Deliver a matching message";
+    my $msg1 = $self->{gen}->generate(
+        subject => "Message 1",
+        extra_headers => [['X-Original-Delivered-To', 'xxx@yyy.zzz'],
+                          ['X-Original-Delivered-To', 'foo@yyy.zzz']],
```

Comment:
One remaining nit: I think this test proves that it _will_ match on the _first_ header (as we're defining first), but it doesn't prove that it _won't_ match on a _non-first_ header.  If you added a `$msg3` to the test with the same headers but these two reversed, then assert that it winds up in INBOX rather than INBOX.matches, that would give us full coverage of the behaviour we want aiui.

Reply:
Something like this seems to work:

```
diff --git a/cassandane/tiny-tests/Sieve/jmapquery_multiple_deliveredto b/cassandane/tiny-tests/Sieve/jmapquery_multiple_deliveredto
index 276b65270..4e4111f6e 100644
--- a/cassandane/tiny-tests/Sieve/jmapquery_multiple_deliveredto
+++ b/cassandane/tiny-tests/Sieve/jmapquery_multiple_deliveredto
@@ -68,11 +68,25 @@ EOF
     $self->check_messages({ 1 => $msg1 }, check_guid => 0);
 
     xlog $self, "Deliver a non-matching message";
-    my $msg2 = $self->{gen}->generate(subject => "Message 2");
+    my $msg2 = $self->{gen}->generate(
+        subject => "Message 2",
+        extra_headers => [['X-Original-Delivered-To', 'foo@yyy.zzz'],
+                          ['X-Original-Delivered-To', 'xxx@yyy.zzz']
+                          ],
+        mime_type => "multipart/mixed",
+        mime_boundary => "047d7b33dd729737fe04d3bde348",
+        body => $body,
+    );
     $self->{instance}->deliver($msg2);
     $msg2->set_attribute(uid => 1);
 
-    xlog "Assert that message got moved into INBOX";
+    xlog $self, "Deliver another non-matching message";
+    my $msg3 = $self->{gen}->generate(subject => "Message 3");
+    $self->{instance}->deliver($msg3);
+    $msg3->set_attribute(uid => 2);
+
+    xlog "Assert that messages got moved into INBOX";
     $self->{store}->set_folder('INBOX');
-    $self->check_messages({ 1 => $msg2 }, check_guid => 0);
+    $self->{store}->set_fetch_attributes('uid');
+    $self->check_messages({ 1 => $msg2, 2 => $msg3 }, check_guid => 0);
 }
```

Reply:
Done